src/Core/Services/Implementations/OrganizationService.cs

```diff
@@ -1110,10 +1110,32 @@ private async Task SendInviteAsync(OrganizationUser orgUser, Organization organi
                 throw new BadRequestException("Invalid token.");
             }
 
+            ICollection<Policy> orgPolicies = null;
```

Comment:
`Lazy<T>` doesn't seem to support async operations. I found a few blogs and the like suggesting things like [creating a class and manually managing the tasks](https://devblogs.microsoft.com/pfxteam/asynclazyt/), but I'm not sure if that level of investment is worth it here, since it would ultimately result in more lines of code than just checking for a null value. I'm not very familiar with the Lazy<T> class, though (thanks for the TIL), so maybe I'm just missing something.

Another option would be to do something like this:

```
ICollection<Policy> orgPolicies = null;
var userOrgs = await _organizationUserRepository.GetManyByUserAsync(user.Id);
var checkOnlyOrg = userOrgs.Any(ou => ou.OrganizationId != orgUser.OrganizationId && ou.Status != OrganizationUserStatusType.Invited);
var checkTwoFactor = !await userService.TwoFactorIsEnabledAsync(user);
if (checkOnlyOrg || checkTwoFactor)
{
    orgPolicies = await _policyRepository.GetManyByOrganizationIdAsync(orgUser.OrganizationId);
};

if (checkOnlyOrg)
{   
    if (orgPolicies.Any(policy => policy.Type == PolicyType.OnlyOrg && policy.Enabled))
    {
        throw new BadRequestException("You cannot join this organization until you are not a " +
            "part of any other organizations.");
    }
    var userPolicies = await _policyRepository.GetManyByUserIdAsync(user.Id);
    if (userPolicies.Any(policy => policy.Type == PolicyType.OnlyOrg))
    {
        throw new BadRequestException("You cannot join this organization because you are already " +
            "member of another organization that restricts having multiple organizations.");
    }
}

if (checkTwoFactor)
{
    if (orgPolicies.Any(p => p.Type == PolicyType.TwoFactorAuthentication && p.Enabled))
    {
        throw new BadRequestException("You cannot join this organization until you enable " +
            "two-step login on your user account.");
    }
}
```

Reply:
that could work, or just a lightweight lazyish implementation with an in-method function:

```c#
// inside method ...
ICollection<Policy> orgPolicies = null;
async Task<bool> hasPolicyAsync(PolicyType policyType)
{
    if (orgPolicies == null)
    {
        orgPolicies = await _policyRepository.GetManyByOrganizationIdAsync(orgUser.OrganizationId);
    }
    return orgPolicies.Any(p => p.Type == policyType && p.Enabled);
}

// ... more fun stuff
if (userOrgs.Any(ou => ou.OrganizationId != orgUser.OrganizationId && ou.Status != OrganizationUserStatusType.Invited))
{
    if (await hasPolicyAsync(PolicyType.OnlyOrg))
    {
        throw new BadRequestException("...");
    }
}

// ... etc.
```

Reply:
@cscharf that's sleek. I implemented based on your suggested, but added some fluff to it so I could also check userPolicies with the same method.